### PR TITLE
Check EC point is on curve at Verifier instantiation

### LINKED ISSUE #115: missing EC public key validation when creating a new Verifier
Currently, when we instantiate a Verifier we do not check that the supplied public key EC point is on the supposed curve.

This is not best practice.  See for example [BCP225](https://www.rfc-editor.org/rfc/rfc8725.html#section-3.4) in the JWS/JWT context:

> "Some cryptographic operations [...] take inputs that may contain invalid values. This includes points not on the specified elliptic curve or other invalid points (e.g., [Valenta], Section 7.1). The JWS/JWE library itself must validate these inputs before using them, or it must use underlying cryptographic libraries that do so (or both!)."

It'd also result in a `panic` when the go-cose application is built against go1.19.  See go1.19 [release notes](https://tip.golang.org/doc/go1.19):

> "Operating on invalid curve points (those for which the IsOnCurve method returns false, and which are never returned by Unmarshal or by a Curve method operating on a valid point) has always been undefined behavior and can lead to key recovery attacks. If an invalid point is supplied to Marshal, MarshalCompressed, Add, Double, or ScalarMult, they will now panic."

This can potentially widen the DoS surface of a go-cose application.  Users should be shielded from such situation.

### PR DESCRIPTION
**Pull request by thomas-fossati**
Fix #115

Signed-off-by: Thomas Fossati <thomas.fossati@arm.com>

### CODE REVIEW

File: verifier_test.go [hunk 1]
```diff
@@ -88,6 +111,12 @@ func TestNewVerifier(t *testing.T) {
 			alg:     0,
 			wantErr: true,
 		},
+		{
```
Comment (yogeshbdeshpande) [inline verifier_test.go:114]: Just a question on the overall method which is verified in this logic?

Should we not compare the exact error returned from the Verifier with the expected error which we are looking for.

It seems we only check whether we "wantErr" or not?

**Review by qmuntal — APPROVED** [review]

**Comment by thomas-fossati** [comment]
@shizhMSFT is it OK to merge this?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -9,7 +9,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        go-version: [1.18, 1.19]
+        go-version: [1.17, 1.18, 1.19]
```
Comment (shizhMSFT) [inline .github/workflows/ci.yml:12]: Question: Do we still support `go 1.17`? I think the min supported version is `go 1.18`.

File: verifier_test.go (hunk 1 above)
Comment (shizhMSFT) [reply verifier_test.go:114]: It seems that we don't care about the detailed errors but has error or nil error.

File: verifier.go [hunk 3]
```diff
@@ -46,6 +46,9 @@ func NewVerifier(alg Algorithm, key crypto.PublicKey) (Verifier, error) {
 		if !ok {
 			return nil, fmt.Errorf("%v: %w", alg, ErrAlgorithmMismatch)
 		}
+		if !vk.Curve.IsOnCurve(vk.X, vk.Y) {
```
Comment (shizhMSFT) [inline verifier.go:49]: Checking only the `IsOnCurve` is OK for `go 1.18` or above. But not for go `1.17` (see [CVE-2022-23806](https://github.com/advisories/GHSA-8c83-vp4v-h7fq)).

Details see https://github.com/golang/go/issues/50974.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (thomas-fossati) [reply .github/workflows/ci.yml:12]: @qmuntal do you remember what was the rationale for also checking 1.17?  My memory got erased in the meantime :-)

File: verifier.go (hunk 3 above)
Comment (thomas-fossati) [reply verifier.go:49]: I didn't know, thanks for pointing that out.

File: verifier_test.go (hunk 1 above)
Comment (thomas-fossati) [reply verifier_test.go:114]: sure, that's how it is at the moment.  Yogesh has a point though that we could improve the precision of the checks.

Comment (yogeshbdeshpande) [reply verifier_test.go:114]: I will raise a separate issue to track this,

Comment (yogeshbdeshpande) [reply verifier_test.go:114]: Raised issue #118

File: .github/workflows/ci.yml (hunk 2 above)
Comment (qmuntal) [reply .github/workflows/ci.yml:12]: I was concerned about the go1.17 behavior commented by @shizhMSFT in https://github.com/veraison/go-cose/pull/116#discussion_r1032556161.

Comment (thomas-fossati) [reply .github/workflows/ci.yml:12]: thanks Quim.  I propose we drop 1.17 from the test matrix and make explicit in the README that we currently support 1.18 and 1.19.

@SteveLasker @roywill WDYT?

Comment (thomas-fossati) [reply .github/workflows/ci.yml:12]: gently prodding the @veraison/go-cose-maintainers ☝️

Comment (shizhMSFT) [reply .github/workflows/ci.yml:12]: I would suggest keeping `go 1.17` dropped.

Comment (SteveLasker) [reply .github/workflows/ci.yml:12]: To minimize the support matrix, I'd vote for dropping and adding if someone opens a PR needing support, with a good explanation.

Comment (thomas-fossati) [reply .github/workflows/ci.yml]: @shizhMSFT @SteveLasker thanks for your input.  I have reverted the support matrix addition in [dd75bac](https://github.com/veraison/go-cose/pull/116/commits/dd75bac21583d93ea0f4f4c3bd6281ed0e5f083d)

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**